app/models/test_suite.py

```diff
@@ -28,43 +28,61 @@ class TestSuiteDocument(modb.BaseDocument):
     A test suite is a document that can store test cases and/or test sets ran.
     """
 
-    def __init__(self, name, lab_name, build_id=None, version="1.0"):
+    def __init__(self, name, lab_name):
         """The test suite document.
 
         :param name: The name given to this test suite.
         :type name: string
         :param lab_name: The name of the lab running this test suite.
         :type lab_name: str
-        :param build_id: The ID of the defconfig associated with this
-        test suite.
-        :type build_id: str
-        :param version: The version of the JSON schema of this test suite.
-        :type version: str
         """
-        self._created_on = None
-        self._id = None
         self._name = name
-        self._version = version
-
-        self.build_id = build_id
-        self.lab_name = lab_name
+        self._lab_name = lab_name
+        self._build_id = None
+        self._version = None
+        self._id = None
+        self._created_on = None
 
         self.arch = None
         self.board = None
         self.board_instance = None
-        self.boot_id = None
+        self.boot_log = None
+        self.boot_log_html = None
```

Comment:
This looks like a lot of duplicates, why not creating a parent document such as "test_environment". That would hold all the info relative to the board / testing environment. After all, all these fields would be identical for a given serie of tests, that would be a lot of wasted space in the DB.

The backend could create one element of test_environment for each callback to test. Then all the test suites would be childs of this parent element.
 
I'm thinking about these fields specifically:
- _lab_name
- _build_id
- arch
- board
- boot_log
- boot_log_html
- boot_result_description
- compiler
- compiler_version
- compiler_version_full
- cross_compile
- defconfig
- defconfig_full
- device_type
- dtb
- dtb_addr
- endian
- file_server_resource
- file_server_url
- initrd_addr
- job
- job_id
- kernel
- kernel_image
- kernel_image_size
- load_addr
- mach
- qemu
- qemu_command

Reply:
Oh, I thought this information was populated only once for each LAVA job in a "test suite" entry, so it shouldn't be duplicated.  Each test case in the suite then has a oid pointing to the suite.  Of course this is also in the boot entries, but that's normal as they're decoupled now.

The "test environment" is essentially a different combination of all these parameters for each LAVA job.  To save a bit of space, we could point to the build oid rather than copy fields from it (compiler, kernel_image, dtb, endian, defconfig).  Then the db queries would need to be adjusted but that seems pretty straightforward.

Might be worth some diagram.  Something like this essentially:

build
-----
* arch
* compiler
* dtb
* git_describe
* ...

test_suite
----------
* oid -> build
* name
* lab_name
* device_type
* boot_log
* ...

test_case
---------
* oid -> test_suite
* status
* time
...

Would this make any sense?

Reply:
Yes makes sense. I didn't know what was your previous issue with boot/test duplicates. That's why I proposed to have a new document. But if it can just point to a build doc it's even better. 

I guess we could only keep lab_name and device_type in test suite. The dtb is tied to the build anyway ? 

Boot_log could be renamed test_suite_log (not necessarily now but in a future PR). And we could keep boot_log for the lava test suite ?

Essentially, one lava job reports one lava test_suite and potentially several other test_suites. So there is multiple test suites for each callback. Then it uses a structure to report the test elements which corresponds to the lava test structure: set and case.

Reply:
Ah OK, I think I didn't appreciate the part about multiple test suites in a single job as I've only been running simple jobs with only one test definition.  I'll add more test cases to my job template.  Still I should already have 2 suites: "lava" and a "simple" one with my current one - will take a closer look at this part.

Any further changes to the tables should probably be done in additional patches so we can see what differs from this version, which just mechanically adds fields from boot to test entries.

The dtb is actually part of the LAVA job data, not the build.  Yes the dtb files are coming from the build job, but it produces all the dtbs it can.  Then when submitting a job, only one dtb is picked (if applicable) for the target platform.  This is the piece of information we get from the callback.

About the log, well I just kept the same name as in boot entries i.e. `boot_log`.  Maybe we should just call it `log`?  The table name should make it obvious what kind of log it is, just like the `name` fields.

So I see what you mean about not duplicating data in several test suite entries.  Now, I think the common data should be in a different document.  We can't call it "job" as that's already used to call a category of tests (as shown on the front-end).  So here's what I mean, with a new "test" table which should have one entry per LAVA job (not strictly tied to LAVA though):

build
---
* arch
* compiler
* git_describe
* defconfig
...

test
---
* oid -> build
* lab_name
* device_type
* log
...

test_suite
---
* oid -> test
* name
...

test_case
---
* oid -> test_suite
* name
* status
...

Sorry this is getting a bit long...  I believe this should completely avoid duplication of data, i.e. things parsed from the callback should only be stored in one field in one table in the database.  It will probably make queries on the API front a bit indirect but I don't think this should be an issue; would be good to measure the performance impact though.

@MiloCasagrande Do you think it's a good idea to segment the data and avoid duplicates as in the scheme above?

Reply:
Full ack on your latest comment @gctucker if we could start like that and send PRs to further changes that is good for me :)

Reply:
Usually I would advise against modeling the data in that way, but it depends.

The main user of the API and the data is going to be the frontend (but not only), so if it takes a lot of queries to retrieve the minimum data necessary for the frontend it's a no go. With data modeled in that way it becomes hard to answers questions like:
- What are the test cases associated with this build?
- What are the test cases associated with this tree?
- What are all the tests run by a single lab?

These are things that could be added to the frontend to extend it, or queries that we could run to do data analysis. With that data model, the backend would have to traverse at least 3 collections to retrieve the test cases ran (with a top-down approach even more with a bottom-up). If you embed or duplicate the data you have to traverse 1 collection. 

Although it might seem an anti-pattern, duplication or denormalization of data is actually welcomed in mongodb, and thinking about normalizing the data is an actual anti-pattern. The best approach when modeling the data with mongodb is: there shouldn't be any relations between collections, if there are, the data must be duplicated.

It's almost impossibile to not see relations between the various collections, but they should be kept at a minimum. It's possible to support queries like the questions above, but you might end up having to write really custom ones for each of the collections.

I wouldn't be concerned with space efficiency under this circumstances: it's not that we are duplicating massive amounts of data like parsed logs that can be megabytes in size . Documents are capped at 16MB, and our average document size is 1.9Kb (even the parsed logs summary average document size is way below the maximum level, and we are at 45Kb).

To support those kind of data layout, with documents linked through their `_id`, mongodb has a JOIN-like operator that performs a left-outer-join, but using it has its own costs: it's reliable and it's fast if you have massive indexes or if your data set is trivial, but the aggregation operations are not always that straightforward to write and the more collections you need to cross the harder it gets.

To be honest, I'm not so sure trying to normalize the data here is a good solution. Maybe some of the fields are not all that necessary in all the collections, but one of the thing discussed in the past was that we should have added more fields into the `test_set` and `test_case` collections to support more detailed queries (like on `job`, `kernel`, `defconfig`... values).

Reply:
@MiloCasagrande OK this is what I was wondering about, as mongodb is not a relational database with things linked to each other everywhere.  I was not suggesting that this should be exposed on the API though, but that the backend would put together the data with JOIN-like queries to return a flat structure to the client in one API call.

Found this part of the "little mongodb book" particularly relevant here:
https://github.com/karlseguin/the-little-mongodb-book/blob/master/en/mongodb.markdown#denormalization

Now it feels like we need a list of pros/cons reasons to make a good decision about splitting the tables or copying the data...  And in any case, we should review which fields we actually care about.  I already did a bit of this with the BootDocument, I only copied the ones that appeared to be actually used anywhere in the backend or frontend into the TestSuiteDocument with this PR.

Reply:
@MiloCasagrande thanks for your input. I must admit, your comment leaves me a bit puzzled.
I am not a front end, database developer so I might miss some basic concepts for you guys.

But when I started working on the test documents structure in the database; I noticed that they are organized in a structured way: case belongs to sets that belongs to suites. 
If one would want to see in the frontend:
- all the test cases belonging to a test suite and test set.
He would need to find all the case and sets IDs in the suite document, and then search for this IDs in the DB. So it's already a three query approach where you could only do one if all the elements where in one single document.

Starting from this structure I extended it with more data. Now that we want to decouple boots from tests I thought it makes sense to create a "test" document that would have references to the test suites. This would be the exact same architecture than suites, sets and cases... 
Hence why I am a bit lost. What differs now ? You are suggesting that this isn't the correct approach but this approach was used to model the test structure at first. Why change now ?

Also, I am not against your approach, if duplicating data doesn't have a drawback on performance or space, why not just creating one big test document that contains everything ? This document would have all the test information from the test platform, build, names to all it's test suites, test sets and test cases results embedded. It would also be a lot easier on the frontend as there would be just one query on one big test document....

Also, using todays DB and proposed architecture, to answer your questions:
- What are the test cases associated with this build?
  - Now: From the build find all boots; from boots find all test suites; from test suites find all test cases
    - 3 queries
  - Proposition: From build find test; from test find suites; from suites find cases
    - 3 queries
- What are the test cases associated with this tree?
  - Now: search in build; from build find boots; from boot find suites; from suite find cases
    - 4 queries
  - Proposition: search in build; from build find test; from test find suites; find cases
    - 4 queries
Or maybe I am missing something and it is much easier today ? 

My point being:
- if it is denormalized : why not just create one big test document ?
- if there is some kind of normalization: why not replacing the boot document by a "test" document containing the same kind of information ?


Also an extract from @gctucker link on the "little monodb book":
_This doesn't mean you should duplicate every piece of information in every document. However, rather than letting fear of duplicate data drive your design decisions, consider modeling your data based on **what information belongs to what document.**_

The question is: what information belongs to what document ? :) 
IMHO I think that the test environment doesn't belong to a test suite, set or case that will be run on multiple platforms (and hence different test env). But I might be wrong, if you could enlighten me on that I'd be thankful !